### PR TITLE
Support sushi-ignoreWarnings.txt file to ignore warnings

### PR DESCRIPTION
**Pull request by ngfreiter**
This addresses [CIMPL-787](https://standardhealthrecord.atlassian.net/browse/CIMPL-787) by adding support for a `sushi-ignoreWarnings.txt` file which can list warnings to be ignored either by directly listing the warning, or by listing a regex that matches the desired warning.

The file must be named `sushi-ignoreWarnings.txt` and must be in the root directory that SUSHI is working on. Only warnings can be ignored, and each line of the file should either list the full content of a warning to be ignored, for example:
```
Instance I2 is not an instance of a resource, so it should only be used inline on other instances, and it will not be exported to a standalone file. Specify "Usage: #inline" to remove this warning.
```
or list a regex that will be used to match certain messages:
```
/Detected the following non-conformant Resource definitions.*/
```

The only way to ignore a multi-line warning is with a regex.

To test this, I would recommend using a FSH file that will generate some warnings. Here is the sample I was using: https://fshschool.org/FSHOnline/#/share/3flTq4D. In `sushi-beta.2` that should generate 3 warnings, one of which is split over multiple lines. Then test out different ways of ignoring those messages, and ensure it aligns with your expectations.

This will need to be documented on FSHSchool, I'll make a corresponding PR there and link it here when I'm done with it, but first I'm going to let people do an initial review in case anything major changes.

### CODE REVIEW

**Review by cmoesel — CHANGES_REQUESTED** [review]
It felt like the bits in the logger are more complex than they need to be (w/ `suppressLogger` and `loggerIsSuppressed` variables that act kind of like globals), so I dug in a little.

According to the docs [here](https://github.com/winstonjs/winston#filtering-info-objects), it seems that if a formatter returns `false`, then that `info` object will not get logged at all.  Unless I'm misunderstanding it, it seems that your `ignoreWarnings` formatter can just return false for ignored warnings and then you don't need to do anything else.  Would you agree or am I missing something?

File: src/utils/FSHLogger.ts [hunk 1]
```diff
@@ -72,6 +95,10 @@ const trackErrorsAndWarnings = format(info => {
 });
 
 const printer = printf(info => {
+  if (loggerIsSuppressed) {
+    logger.transports[0].silent = false;
```
[the remark this answers is not in the document]
Comment (cmoesel) [reply src/utils/FSHLogger.ts:99]: Actually, if the `ignoreWarnings` formatter can just return `false` then this bit of code is irrelevant I guess.

[the remark this answers is not in the document]
Comment (ngfreiter) [reply src/utils/FSHLogger.ts]: Updated `ignoreWarnings` to just return `false` in https://github.com/FHIR/sushi/pull/894/commits/eed8c613ae01f5dffbcd7c17db6d60c4c7059ebd.

File: src/utils/FSHLogger.ts [hunk 2]
```diff
@@ -30,6 +31,28 @@ const withLocation = format(info => {
...
+  }
+
+  const shouldIgnore = ignoredWarnings?.some(m => {
+    if (m.startsWith('/') && m.endsWith('/')) {
+      const regex = new RegExp(m.slice(1, -1));
+      return regex.test(info.message);
+    } else {
+      return m === info.message;
+    }
+  });
```
[the remark this answers is not in the document]
Comment (ngfreiter) [reply src/utils/FSHLogger.ts]: Updated the regex building in https://github.com/FHIR/sushi/pull/894/commits/eed8c613ae01f5dffbcd7c17db6d60c4c7059ebd.

[the remark this answers is not in the document]
Comment (ngfreiter) [reply src/utils/FSHLogger.ts]: And https://github.com/FHIR/sushi/pull/894/commits/e9a7616a62508400608f7d551bffe45bd0e818c6 makes it so whitespace is trimmed.

**Comment by ngfreiter** [comment]
Thanks for the comments @cmoesel, they make things much cleaner. Should be ready for re-review.

**Comment by cmoesel** [comment]
@ngfreiter -- In FHIR's `ignoreWarnings.txt`, you can use `#` to mark comments explaining why you're ignoring something.  Would we want to support something like that?  For example:
```
# ignore everything because I'm lazy
/.*/
```

This technically works already, because we just treat the first line like it is a message to be ignored -- and it just doesn't match any messages.  But it seems like if we do want this as a feature, maybe it would be better to filter out lines starting w/ `#` when we're collecting the ignored warnings.  What do you think?

**Comment by ngfreiter** [comment]
> @ngfreiter -- In FHIR's `ignoreWarnings.txt`, you can use `#` to mark comments explaining why you're ignoring something. Would we want to support something like that?

I knew I was forgetting something. Added the comments in https://github.com/FHIR/sushi/pull/894/commits/bd20a93d2333fdbb0a25767e6b0c1d51cfdfed6b.

**Review by jafeltra — COMMENTED** [review]
This functionally looks really good! I just had one question about the location of the file. I think it makes sense to keep in the root folder since that's where the sushi config file is, but I did initially get confused and wanted to put it next to the publisher's ignoreWarnings files inside of `input`. Is there any benefit to putting our file in the same location as the publisher's is expected to be? Semantically, I think that it makes more sense to keep ours where it is, but I wanted to run this same thought process past you.

**Comment by cmoesel** [comment]
We could be flexible and allow for it in either place.  If it's in both places, we could emit a warning and then either choose one (root) or concatenate them together so it ignores anything in either file.

**Comment by jafeltra** [comment]
> We could be flexible and allow for it in either place. If it's in both places, we could emit a warning and then either choose one (root) or concatenate them together so it ignores anything in either file.

But can you ignore that warning? 😆 

More seriously, I did think about that option, but I do almost feel a little nervous about supporting two ways to do something after the ig-data/folder structure fun. But this is fairly limited and would be helpful probably so I'm not opposed.

**Comment by ngfreiter** [comment]
> This functionally looks really good! I just had one question about the location of the file. I think it makes sense to keep in the root folder since that's where the sushi config file is, but I did initially get confused and wanted to put it next to the publisher's ignoreWarnings files inside of `input`. Is there any benefit to putting our file in the same location as the publisher's is expected to be? Semantically, I think that it makes more sense to keep ours where it is, but I wanted to run this same thought process past you.

Updated it so that the file can be placed in both locations in https://github.com/FHIR/sushi/pull/894/commits/d185d0affc1074490ff41e99e007f40b4cc2f989. If it is in only one location, that file is used. If the file is in the `input` folder and the root, only the one in the root is used, and a warning is logged. And yes, you can ignore the warning about `ignoreWarnings` 🙄 .

**Comment by ngfreiter** [comment]
If you review this PR, please also give the documentation a look: https://github.com/FSHSchool/site/pull/45

**Review by jafeltra — APPROVED** [review]
This looks good to me! Thanks for letting me ignore the ignoreWarnings warning.

**Review by cmoesel — APPROVED** [review]
Yep!  Looks good and works well!

**Pull request merged**